### PR TITLE
Don't synchronize new browser created during drag-and-drop

### PR DESCRIPTION
**Pull request by sangwoo108**
We were synchronizing the new browser created during drag-and-drop and it caused the new browser to be dangling in some cases.

For e.g.
* Drag a non-pinned tab out of a window to create a new window
* Don't drop and drag it back to the original window
* Open tab search button and see if the dummy pinned tab is gone. => The dummy tab for new window can be dangling, even the browser looks gone.

Resolves https://github.com/brave/brave-browser/issues/38764

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

* Drag a non-pinned tab out of a window to create a new window
* Don't drop and drag it back to the original window
* Open tab search button and see if the dummy pinned tab is gone. 
  => Previously: The dummy tab for new window can be dangling, even the browser looks gone.
  => Expected: There shouldn't be no dangling web page

### CODE REVIEW

File: browser/ui/tabs/shared_pinned_tab_service.cc [hunk 1]
```diff
@@ -926,3 +948,7 @@ SharedPinnedTabService::CreateDummyWebContents(
                                           shared_contents);
   return dummy_contents;
 }
+
+bool SharedPinnedTabService::IsBrowserInTabDragging(Browser* browser) const {
+  return static_cast<BraveBrowserWindow*>(browser->window())->IsInTabDragging();
```
Comment (iefremov) [inline browser/ui/tabs/shared_pinned_tab_service.cc:953]: `CHECK(browser);` ?

File: browser/ui/tabs/shared_pinned_tab_service.cc [hunk 2]
```diff
@@ -721,7 +722,24 @@ void SharedPinnedTabService::SynchronizeMovedPinnedTab(int from, int to) {
   }
 }
 
+void SharedPinnedTabService::TabDraggingEnded(Browser* browser) {
+  if (!in_tab_dragging_browsers_.count(browser)) {
+    return;
+  }
+
+  in_tab_dragging_browsers_.erase(browser);
```
Comment (iefremov) [inline browser/ui/tabs/shared_pinned_tab_service.cc:730]: you can simply call `erase` (and don't call `count` above) and check it's return value to see if anything was erased

**Review by iefremov — APPROVED** [review]

File: browser/ui/views/frame/brave_browser_frame.cc [hunk 3]
```diff
@@ -58,3 +62,16 @@ views::internal::RootView* BraveBrowserFrame::CreateRootView() {
...
   return root_view_;
 }
+
+void BraveBrowserFrame::SetTabDragKind(TabDragKind kind) {
+  const bool should_sync_shared_pinned_tab =
+      base::FeatureList::IsEnabled(tabs::features::kBraveSharedPinnedTabs) &&
+      tab_drag_kind_ == TabDragKind::kAllTabs && kind == TabDragKind::kNone;
+
+  BrowserFrame::SetTabDragKind(kind);
+
```
Comment (simonhong) [inline browser/ui/views/frame/brave_browser_frame.cc:72]: nit:
```suggestion
  BrowserFrame::SetTabDragKind(kind);

  const bool should_sync_shared_pinned_tab =
      base::FeatureList::IsEnabled(tabs::features::kBraveSharedPinnedTabs) &&
      tab_drag_kind_ == TabDragKind::kAllTabs && kind == TabDragKind::kNone;

```

**Review by simonhong — COMMENTED** [review]
btw, shared pinned tab works?
Pinned tab is not shared on my local build and also niglty on macOS. (enabled from brave://flags)

**Comment by sangwoo108** [comment]
> Pinned tab is not shared on my local build and also niglty on macOS. (enabled from brave://flags)

There's a preference now

Apperaence > Tabs

image

File: browser/ui/tabs/shared_pinned_tab_service.cc (hunk 2 above)
Comment (sangwoo108) [reply browser/ui/tabs/shared_pinned_tab_service.cc]: Nice catch! Done.

File: browser/ui/views/frame/brave_browser_frame.cc (hunk 3 above)
Comment (sangwoo108) [reply browser/ui/views/frame/brave_browser_frame.cc:72]: Shouldn't we check the old value(tab_drag_kind_) before calling Browser::SetTabDragKind(), which sets the new value to the member variable?

Comment (simonhong) [reply browser/ui/views/frame/brave_browser_frame.cc:72]: Ah, I see. it's set by `BrowserFrame::SetTabDragKind()`.
Just suggested to move vars closely where it's used. but it shouldn't in this case.

**Comment by simonhong** [comment]
> > Pinned tab is not shared on my local build and also niglty on macOS. (enabled from brave://flags)
> 
> There's a preference now
> 
> Apperaence > Tabs
> 
> image

Oops, forgot about it :)

**Review by simonhong — APPROVED** [review]
++

**Pull request merged**